### DISCUSSION TITLE
Custom index examples

### DISCUSSION THREAD

**Discussion by benbovy** [pydata/xarray#7041] [Ideas]
Let's gather here examples (prototypes) of custom indexes that have been proposed so far. This is all very much WIP, some examples may rely on features that have not yet landed in the `main` branch while other examples may be outdated. Anyway, it should provide a good basis for more experiments with Xarray indexes. To @pydata/xarray or anyone: feel free to edit this comment and add to the list below.

EDIT: there's now an index gallery at https://xarray-indexes.readthedocs.io/ - https://github.com/xarray-contrib/xarray-indexes

- default indexes (for reference): [PandasIndex](https://github.com/pydata/xarray/blob/63ba862d03c8d0cd8b44d2071bc360e9fed4519d/xarray/core/indexes.py#L217) and [PandasMultiIndex](https://github.com/pydata/xarray/blob/63ba862d03c8d0cd8b44d2071bc360e9fed4519d/xarray/core/indexes.py#L537)
- Helper class to implement a custom meta-index based on pandas index(es): [notebook](https://notebooksharing.space/view/3d599addf8bd6b06a6acc241453da95e28c61dea4281ecd194fbe8464c9b296f#displayOptions=)
- basic array index: [notebook](https://notebooksharing.space/view/48ad86aed90f7588c9a475be6747528d87f975cb3317e5bd94265ffaa5a2478f#displayOptions=)
- kd-tree index: [notebook](https://github.com/keewis/xarray-custom-indexes/blob/main/custom-index.ipynb)
- unit-aware (pint) index: https://github.com/xarray-contrib/pint-xarray/pull/163 
- periodic boundary index: https://github.com/pydata/xarray/issues/7031
- interval index: https://github.com/pydata/xarray/discussions/6783#discussioncomment-3149033
- functional index: [notebook](https://notebooksharing.space/view/00e4f9cf885fd4624de2eb3a26b779765fb3fbd57e7cf75acd176752064fa613#displayOptions=)
- CRS index (raster 2d): [notebook](https://github.com/dcherian/crsindex/blob/main/crsindex.ipynb) and https://github.com/corteva/rioxarray/issues/588
- out-of-core (dask) index: https://github.com/pydata/xarray/issues/1650 / https://github.com/pydata/xarray/issues/1094 / [notebook](https://notebooksharing.space/view/2814ca3e1498043574b8357d1e77a587b201cdfa50bcaae5d3056b08a9e32fd3#displayOptions=)
- Staggered grid (xGCM) index: https://github.com/dcherian/xindexes/blob/main/xgcmindex.ipynb; https://github.com/xgcm/xgcm/issues/225#issuecomment-763071395
- Shapely (2.0) geometry index: https://github.com/xarray-contrib/xvec - [notebook](https://notebooksharing.space/view/3ccb33d20f33393d0f7acffb6504b2e1223deafe8d9d0f34054c8058c97569fd#displayOptions=)
- World Coordinate System (WCS) index (astropy): [notebook](https://notebooksharing.space/view/2e33c4554e5dfe754306515dbb5f223615ca4f0bbbf54bfbf1494b9417e33d14#displayOptions=), https://github.com/sunpy/ndcube/issues/222
- DGGS (Discrete Global Grid System) indexes: https://github.com/xarray-contrib/xdggs
- Regular or range index: [notebook](https://notebooksharing.space/view/37a45e01bbfe69d4242808ef9e88bf3dbcbd297d0221493fc4a4ae77e980e809#displayOptions=), #8473 
- Forecast Model Run Collections (FMRC) - https://rolodex.readthedocs.io/forecast.html

**Comment by SimonHeybrock** [comment]
1. Not sure if this would work implicitly via the `PintMetaIndex` you linked to, but maybe worth mentioning: We would like something like `pandas.IntervalIndex` but *with units*.

2. In the field I work in the automatic alignment that indexes provide (such as automatically returning the intersection from `da1 + da2` with an index-coord for `'x'`) is not really required. Instead it may create (some) overhead, or risk running long computations where the user would actually just have preferred getting an early exception. Therefore, I am playing with the idea of having a `StrictIndex` (probably bad name, just to have something to refer to) which:
   - Is a very light wrapper around a NumPy array or duck array, as we want to be able to do math on this to compute derived coords.
   - Allows for label-based indexing (`sel`), but performance is not optimized for that as few queries are expected.
   - Raises an exception when trying to join with something that is not identical (length, labels, order of labels). Basically, `da1 + da2` should just raise unless `da1.coords['x']` is identical to `da2.coords['x']` in the NumPy-array-equal sense.

**Reply by benbovy** [reply]
@SimonHeybrock your 2nd point is a good idea, it has actually been suggested as an example to put in the docs (https://github.com/pydata/xarray/pull/6975#issuecomment-1239526376). It may be useful to have it implemented Xarray.

**Reply by benbovy** [reply]
Early prototype of such an `ArrayIndex` in this [notebook](https://notebooksharing.space/view/48ad86aed90f7588c9a475be6747528d87f975cb3317e5bd94265ffaa5a2478f#displayOptions=) (I added it to the list above).

**Reply by SimonHeybrock** [reply]
Very nice! I gave this a try with a `pint.Quantity` (instead of a `np.ndarray`) and I got it to work with very minor changes, not specific to Pint. I can now do (no need for @keewis's `PintMetaIndex` in this case!):

```python
da.sel(x=np.array(2.2) * pint.Unit('m'))
```

`da.sel(x=2.2 * pint.Unit('m'))` or `da.sel(x=pint.Quantity(2.2, 'm'))` do not work without further changes, since `xr.core.utils.is_scalar` returns `False`, but I presume this could be solved somehow.

**Reply by SimonHeybrock** [reply]
There is one subtlety I ran into: Using `np.argmax(self.array == label)` in `ArrayIndex.sel`, we will get index 0 on a units mismatch, but we'd actually like an exception. Adding a special check for Pint would not be so nice, and I was wondering if this is something applicable to other duck arrays as well? An ugly hack that works is to use

```def
class ArrayIndex(Index):
  ...
  def sel(self, ...):
    _ = self.array[0] == label  # trigger exception if incompatible
    indexer = np.argmax(self.array == label)
    ...
```
Is there a better and generic way to ensure that a label is "compatible" with a certain index?

**Reply by benbovy** [reply]
Yes, when experimenting with this `ArrayIndex` I've also found the returned index 0 annoying in a few other cases too.

> Is there a better and generic way to ensure that a label is "compatible" with a certain index?

There is a vague suggestion of a protocol [here](https://github.com/pydata/xarray/blob/main/design_notes/flexible_indexes_notes.md#6-coordinate-duck-arrays).

**Reply by keewis** [reply]
in this specific case I'd actually set `force_ndarray_like = True` on the application registry, which would convert the scalar to a 0d array (if you use `pint-xarray` that will happen automatically to fix some other issues)

**Comment by TomNicholas** [comment]
Spotted another custom index experiment in the wild by @Huite - [unstructured grid indexes](https://github.com/Deltares/xugrid/issues/35). (More [context here](https://github.com/UXARRAY/uxarray/discussions/185#discussioncomment-4726859))

**Comment by TomNicholas** [comment]
[Histogram integration in xarray via flox](https://github.com/pydata/xarray/issues/4610#issuecomment-1423144049) would benefit from using `pandas.IntervalIndex` to represent both ends of each histogram bin. I think there are currently some bugs if you try to create an xarray object with an `IntervalIndex`.

**Reply by TomNicholas** [reply]
@benbovy is an `IntervalIndex` something which needs a wrapper similar to your `MultiPandasIndex` example? cc @dcherian 

> Helper class to implement a custom meta-index based on pandas index(es): [notebook](https://notebooksharing.space/view/3d599addf8bd6b06a6acc241453da95e28c61dea4281ecd194fbe8464c9b296f#displayOptions=)

**Reply by dcherian** [reply]
IntervalIndex seems a little hard to support, because it wraps a Pandas ExtensionArray. I tried some things [here](https://github.com/dcherian/xindexes/blob/main/interval-array.ipynb) (sorry not well documented). One other thing to try could be a numpy structured array.

**Reply by benbovy** [reply]
I think one of the best approaches would be to have a custom Xarray `IntervalIndex` that only supports a unique 1-d coordinate.

That said, an extension array adaptor like the one at the end of @dcherian's notebook would be nice to have. As another example,  geopandas uses extension arrays for the geometry columns so this could be useful for other cases like https://github.com/martinfleis/xvec.

Perhaps something like this could work for IntervalIndex:

- Expose a custom `xarray.indexes.IntervalIndex` that wraps a `pd.IntervalIndex`
  - Like for PandasIndex, its corresponding 1-d coordinate wraps the pandas index in a `xarray.core.indexing.PandasIndexingAdapter`
  - It could actually be implemented as a subclass of PandasIndex?
- How to use set_xindex?
  - `obj.set_xindex("x", IntervalIndex)` where "x" is a 1-d coordinate (if it wraps an `pd.arrays.IntervalArray` or a `pd.IntervalIndex` just reuse it, otherwise create a new pandas interval index with default bounds).
  - `obj.set_xindex("x", IntervalIndex, use_bounds_coord="x_bounds")` where `use_bounds_coord` is an option specific to the Xarray IntervalIndex. In this case the "x_bounds" coordinate remains unindexed, it could be manually removed after creating the index.
- Getting back the bounds coordinate, e.g., with something like `obj.assign(x_bounds=obj.xindexes["x"].get_bounds_coord())`

**Comment by DamienIrving** [comment]
I'm very excited about the `rolodex.ForecastIndex` described at: https://xarray-indexes.readthedocs.io/earth/forecast.html

Apologies if I'm missing something very obvious, but how does one install `rolodex` in order to run `import rolodex.forecast`?

From the [rolodex documentation page](https://rolodex.readthedocs.io/forecast.html) I can't find installation instructions or even a GitHub repo corresponding to that documentation page or a rolodex package?

**Reply by kmuehlbauer** [reply]
I think this is the related GitHub repo: https://github.com/dcherian/rolodex

**Reply by dcherian** [reply]
my goal is to move it to https://github.com/abkfenris/xarray_fmrc but haven't had time to do that yet.
